File: TerraObserver/TO.Abstractions/Models/Planets/ICatlikeCodingNoise.cs
using Godot;
using Godot.Abstractions.Bases;
using TO.Domains.Structs.Tiles;

namespace TO.Abstractions.Models.Planets;

/// Copyright (C) 2025 Zhu Xiaohe(aka ZeromaXHe)
/// Author: Zhu XH (ZeromaXHe)
/// Date: 2025-06-06 17:20:06
public interface ICatlikeCodingNoise : IResource
{
    Image? NoiseSourceImage { get; }
    int HashGridSize { get; }
    HexHash[] HashGrid { get; }
    RandomNumberGenerator Rng { get; }
}

File: TerraObserver/TO.Presenters/Abstractions/Models/Planets/ICatlikeCodingNoise.fs
namespace TO.Presenters.Abstractions.Models.Planets

open Godot
open Godot.Abstractions.Bases
open TO.Domains.Structs.Tiles

/// Copyright (C) 2025 Zhu Xiaohe(aka ZeromaXHe)
/// Author: Zhu XH (ZeromaXHe)
/// Date: 2025-06-27 17:55:27
[<Interface>]
type ICatlikeCodingNoise =
    inherit IResource
    abstract NoiseSourceImage: Image
    abstract HashGridSize: int
    abstract HashGrid: HexHash array
    abstract Rng: RandomNumberGenerator

type GetNoiseSourceImage = unit -> Image
type GetHashGridSize = unit -> int
type GetHashGrid = unit -> HexHash array
type GetRng = unit -> RandomNumberGenerator

[<Interface>]
type ICatlikeCodingNoiseQuery =
    abstract GetNoiseSourceImage: GetNoiseSourceImage
    abstract GetHashGridSize: GetHashGridSize
    abstract GetHashGrid: GetHashGrid
    abstract GetRng: GetRng

module CatlikeCodingNoiseQuery =
    let getNoiseSourceImage (this: ICatlikeCodingNoise) = fun () -> this.NoiseSourceImage
    let getHashGridSize (this: ICatlikeCodingNoise) = fun () -> this.HashGridSize
    let getHashGrid (this: ICatlikeCodingNoise) = fun () -> this.HashGrid
    let getRng (this: ICatlikeCodingNoise) = fun () -> this.Rng
